src/Formats/ProtobufSerializer.cpp

```diff
@@ -3353,20 +3353,63 @@ namespace
-                        bool has_fallback_to_array_of_nested_messages = false;
+                        bool nested_all_arrays = false ;
+                        bool nested_some_arrays = false ;
+
                         if (field_descriptor->is_repeated())
                         {
-                            bool has_arrays
-                                = boost::range::find_if(
-                                      nested_data_types, [](const DataTypePtr & dt) { return (dt->getTypeId() == TypeIndex::Array); })
-                                != nested_data_types.end();
-                            if (has_arrays)
-                                has_fallback_to_array_of_nested_messages = true;
+                            nested_all_arrays = true ;
+                            for(const auto & nested_data_type : nested_data_types)
+                            {
+                                if(nested_data_type->getTypeId() == TypeIndex::Array)
+                                    nested_some_arrays = true ;
+                                else
+                                    nested_all_arrays = false ;
+                            }
+                        }
+
+                        /// Since all selected nested columns are of type array, bypass exception flow :
+                        /// Attempt serialization of nested columns as an array of nested messages directly
+                        if (nested_all_arrays)
+                        {
+                            for (DataTypePtr & dt : nested_data_types)
+                                dt = assert_cast<const DataTypeArray &>(*dt).getNestedType();
+
+                            std::vector<size_t> used_column_indices_in_nested;
+                            auto nested_message_serializer = buildMessageSerializerImpl(
+                                nested_column_names.size(),
+                                nested_column_names.data(),
+                                nested_data_types.data(),
+                                *field_descriptor->message_type(),
+                                /* with_length_delimiter = */ false,
+                                google_wrappers_special_treatment,
+                                field_descriptor,
+                                used_column_indices_in_nested,
+                                /* columns_are_reordered_outside = */ true,
+                                /* check_nested_while_filling_missing_columns = */ false);
+
+                            /// `columns_are_reordered_outside` is true because column indices are
+                            /// going to be transformed and then written to the outer message,
+                            /// see add_field_serializer() below.
+
+                            if (nested_message_serializer)
+                            {
+                                std::vector<std::string_view> column_names_used;
```

Comment:
It's exactly the same as the code after the try-error, right? I think it could be wrapped into a lambda function above and be called here and after the try-error

Reply:
I can help with that:)

Reply:
alright, I've given it a try, looking forward to your feedback :pray: